### PR TITLE
executor: fix incorrect handling expr eval results in projection executor

### LINKED ISSUE #16589: Projection executor crashes when expression returns RpnStackNodeVectorValue::Generated value
### What version of TiKV are you using?

Release Version:   8.0.0-alpha
Edition:           Community
Git Commit Hash:   de0215e136f1c6e12a8dc2e4f39c8171fb33b42f
Git Commit Branch: master
UTC Build Time:    2024-03-01 02:09:19
Rust Version:      rustc 1.77.0-nightly (89e2160c4 2023-12-27)

### Steps to reproduce

https://github.com/pingcap/tidb/issues/34993

### PR DESCRIPTION
**Pull request by yibin87**
### What is changed and how it works?

Issue Number: Close #16589 

What's Changed:

```commit-message
Expression's evaluation results can be three types:

1. Scalar: ScalarValue
2. Generated: RpnStackNodeVectorValue::Generated
3. Ref: RpnStackNodeVectorValue
For Ref type, it maintains a physical vector ref and an logical index vector ref, it is expected to only use indexes in logical index vector to access physical vector.
For Generated type, it maintains only a physical vector, all elements can be accessed.
So for a query like select col0, cast(col0 as signed) from t0 where col0 > 100, the 'col0' projection expression will produce Ref which reuses input column's logical index vector, while 'cast(col0 as signed)' will produce Generated.
However, in current projection executor, we will use input column's logical index vector for both two expression results,  causing index out of range for Generated one.
Change to transform Ref to Generated, and recreate logical index vector to [0...N) to fix the issue.

```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: components/tidb_query_expr/src/types/expr_eval.rs [hunk 1]
```diff
@@ -47,6 +46,45 @@ impl<'a> RpnStackNodeVectorValue<'a> {
...
+                        Json => JsonRef,
+                        Enum => EnumRef,
+                        Set => SetRef,
+                    ],
+                    match result_vec_ref {
+                        VectorValue::TT(dest_column) => {
+                            for index in logical_rows {
+                                let src_ref = TT::borrow_vector_value(physical_value);
+                                // TODO: This clone is not necessary.
+                                dest_column.push(src_ref.get_option_ref(*index).map(|x| x.into_owned_value()));
```
Comment (gengliqi) [inline components/tidb_query_expr/src/types/expr_eval.rs:78]:
```suggestion
                                dest_column.push(src_ref.get_option_ref(*index).unwrap());
```

**Review by wshwsh12 — APPROVED** [review]

File: components/tidb_query_expr/src/types/expr_eval.rs (hunk 1 above)
Comment (gengliqi) [reply components/tidb_query_expr/src/types/expr_eval.rs]: My bad. Should change back.

File: components/tidb_query_expr/src/types/expr_eval.rs [hunk 2]
```diff
@@ -47,6 +46,45 @@ impl<'a> RpnStackNodeVectorValue<'a> {
...
+        match self {
+            RpnStackNodeVectorValue::Generated { physical_value } => Ok(physical_value),
+            RpnStackNodeVectorValue::Ref {
+                physical_value,
+                logical_rows,
+                ..
+            } => {
+                // TODO: extract a common util function to do this
+                let mut result_vec = physical_value.clone_empty(logical_rows.len());
+                let result_vec_ref = result_vec.borrow_mut();
```
Comment (gengliqi) [inline components/tidb_query_expr/src/types/expr_eval.rs:60]:
```suggestion
```

File: components/tidb_query_expr/src/types/expr_eval.rs [hunk 3]
```diff
@@ -47,6 +46,45 @@ impl<'a> RpnStackNodeVectorValue<'a> {
...
+                        Real,
+                        Duration,
+                        Decimal,
+                        DateTime,
+                        Bytes => BytesRef,
+                        Json => JsonRef,
+                        Enum => EnumRef,
+                        Set => SetRef,
+                    ],
+                    match result_vec_ref {
```
Comment (gengliqi) [inline components/tidb_query_expr/src/types/expr_eval.rs:73]:
```suggestion
                    match &mut result_vec {
```

File: components/tidb_query_expr/src/types/expr_eval.rs [hunk 4]
```diff
@@ -47,6 +46,45 @@ impl<'a> RpnStackNodeVectorValue<'a> {
...
+                        VectorValue::TT(dest_column) => {
+                            for index in logical_rows {
+                                let src_ref = TT::borrow_vector_value(physical_value);
+                                // TODO: This clone is not necessary.
+                                dest_column.push(src_ref.get_option_ref(*index).unwrap());
```
Comment (gengliqi) [inline components/tidb_query_expr/src/types/expr_eval.rs:78]:
```suggestion
                                dest_column.push(src_ref.get_option_ref(*index).map(|x| x.into_owned_value()));
```

**Review by gengliqi — COMMENTED** [review]
Rest LGTM

**Review by gengliqi — APPROVED** [review]

**Pull request merged**